### DISCUSSION TITLE
Return total tasks number on a get tasks request

### DISCUSSION THREAD

**Discussion by bidoubiwa** [meilisearch/product#635] [Feedback & Feature Proposal]
## Suggestion

Currently when performing a `GET /tasks` request, these are the fields that are returned 

- results
- limit
- from
- next
 
I would like to add a new field `totalTasks`.

This new field would return the total number of tasks that matches the filters.

For example:
Given an instance on which I have: 
- 30 tasks enqueued
- 50 tasks processing
- 100 tasks done/failed

I ask for the tasks with the `processing` and `enqueued` status  (`statuses=enqueued,processing`) with the default limit of 20.

In the response, I'm provided with my paginated selection of 20 tasks based on my filters, and the total tasks of `80` based on these filters as well.

### First situation

Imagine a dashboard on which I want to keep the user up to date on the numbers of tasks that are still being processed or still enqueued.

```
tasks in progress: 80
```

....

```
tasks in progress: 55
```

etc etc until done.

```
tasks in progress: 0
```

Achieving this with the current state of the API is very hard. As I need to paginate on the whole `GET /tasks` and hope that once I'm done with the pagination, no new tasks have been added.
This pagination would have to be done every time I decide to refresh the dashboard with the new state of the tasks.

### Second situation

It would also make it possible to know the total amount of tasks by requesting `GET /tasks` without any filter. This might help the user create an automatic task to delete the tasks when reaching a certain threshold.

**Comment by gmourier** [comment]
Thank you, @bidoubiwa !

cc @meilisearch/cloud-team If you wanna add precisions to the conversation so you don't have to create another discussion!

**Reply by davelarkan** [reply]
Thanks @bidoubiwa !!

Your requirements describe ours exactly.

**Reply by gmourier** [reply]
From @bidoubiwa's request, I understand that only a `totalTasks` integer is needed, while I sense it could be more and we may lack details. Please correct me if I'm wrong, but wouldn't having a stats breakdown per status be better to give a more precise report? 

e.g. GET `/tasks`

```jsonc
{
  "results":[
    {
      "uid":1,
      "indexUid":"movies_reviews",
      "status":"failed",
      "type":"documentAdditionOrUpdate",
      "canceledBy": null,
      "details":{
        "receivedDocuments":100,
        "indexedDocuments":0
      },
      "error": null,
      "duration":null,
      "enqueuedAt":"2021-08-12T10:00:00.000000Z",
      "startedAt":null,
      "finishedAt":null
    },
    {
      "uid":0,
      "indexUid":"movies",
      "status":"succeeded",
      "type":"documentAdditionOrUpdate",
      "canceledBy": null,
      "details":{
        "receivedDocuments":100,
        "indexedDocuments":100
      },
      "error": null,
      "duration":"PT16S",
      "enqueuedAt":"2021-08-11T09:25:53.000000Z",
      "startedAt":"2021-08-11T10:03:00.000000Z",
      "finishedAt":"2021-08-11T10:03:16.000000Z"
    },
    ...
  ],
  "limit": 20,
  "from": 1,
  "next":null,
  "total": 6 //the total number of tasks matching the filters; no filter -> the total number tasks
  "stats": { //a statistics breakdown per task enum matching the filters; no filter -> a stats breakdown of all tasks
      "byStatuses": {
            "enqueued": 0,
            "processing": 0,
            "failed": 1,
            "succeeded": 5,
            "canceled": 0
       }
  }
}
```

**Reply by davelarkan** [reply]
Indeed the stats breakdown would be really nice. 

If that were available we would likely make use of it. 

Having said that, at a minimum if we had the total it would already make our lives so much easier. The stats would just be the cherry on top.

**Reply by irevoire** [reply]
Why not also return a breakdown of the other faceted fields (maybe not the dates);
```json
"stats": {
       "byIndexes": {
            "movies": 1,
            "movies_review": 1,
       },
      "byStatuses": {
            "enqueued": 0,
            "processing": 0,
            "failed": 1,
            "succeeded": 5,
            "canceled": 0
       },
       "byTypes": {
            "documentAdditionOrUpdate": 2,
            "taskDeletion": 1,
            "taskCancelation": 3,
            "dumpCreation": 1,
            "snapshotCreation": 139,
       },
       "canceled": 53, // maybe we could do a more precise breakdown as well; I don't think the breakdown would be useful for anyone here
}
```

Also, I do think that the stats of the tasks is a completely different subject that should be treated in another discussion.

**Reply by gmourier** [reply]
I like your proposal @irevoire; `canceled` is already described in`byStatuses`; is there a difference with the one described at the root level of the `stats` object?

**Reply by irevoire** [reply]
Ah yeah true sorry, we don't need it :+1:

**Reply by bidoubiwa** [reply]
I still like `totalTasks` for whatever mix of filters I throw. It's not only based on one specific facet, I might want all `processed` from a certain `index` in which case the `stats` does not provide that information (unless I misunderstood).

Same regarding the dates, I might want the total tasks in a date range.

**Comment by irevoire** [comment]
Should `totalTasks` be impacted by the `from` parameter?
For example, if I sent `/tasks?from=0`, should `totalTasks` return `1` forever? Or should it increase every time a new task is enqueued?

Sub question whatever the answer to the first one, I think it would be nice if there was a way to get both information:
- The number of tasks remaining AFTER the `from`
- The total number of tasks that matches the filter without the `from` and the `filter`

**Reply by gmourier** [reply]
Is `totalTasks` still needed if the `stats` breakdown exists?

> For example, if I sent /tasks?from=0, should totalTasks return 1 forever?

I would say no; from what I understand the purpose is to know the total number of tasks matching the filters among multiple "page"

**Reply by bidoubiwa** [reply]
`totalTasks` for me should only be based on the `filters` field. Having the number of tasks after a certain from is certainly a good thing to consider for a smooth finite pagination experience.

> Is totalTasks still needed if the stats breakdown exists

For me yes, see [this comment](https://github.com/meilisearch/product/discussions/635#discussioncomment-5904323)

**Reply by irevoire** [reply]
The filters could be applied before returning the stats as well, but then once again, it's inconvenient and hard to read; filter or not, it should be on another route to me while the `totalTasks` don't pollute much and can be helpful when you're looking for specific information.

I just spent this morning trying to debug the instance of a user; I can tell you that `totalTasks` would have been helpful like 10 times while the complete stats would never have been useful and even super boring (it's already hard to find the info you want on the task queue, let's not add 50 lines of useless stats in it please I beg you)

**Reply by gmourier** [reply]
I'm all for having the `stats` in the `/tasks` routes. At least, they can be added by a parameter like `getStats`: true/false, if it's too much by default for a  human reader (still it will be located at the end of the response).

If you want to build an interface to navigate the tasks there is a huge chance that you will want to display some counter next to the different possible filters (by indexes, by statuses, by types). Not injecting the stats in the same route will make it a bit more complicated because 2 calls have to be managed.

> of useless stats in it

If it's useless we should not add anything at all 🥸

**Reply by irevoire** [reply]
Hiding it by default with a param to get it could be an acceptable solution; yes:+1: 
But personally, I would still prefer a new route with a clear distinction between
- Getting stats on the task queue
- Navigating the task queue

> If you want to build an interface to navigate the tasks there is a huge chance that you will want to display some counter next to the different possible filters (by indexes, by statuses, by types). Not injecting the stats in the same route will make it a bit more complicated because 2 calls have to be managed.

tbh that's the only use case where it makes sense to mix both features to me, and I'm not even sure it would work in practice :thinking: 
What should happen if the user scrolled a lot? Should the stats only take into account the tasks after the `from`?
Same with any kind of filter, basically, as @bidoubiwa was saying, should the stats take the filter into account?
Both answers look wrong to me and depend on how you want to build your UI, so I would say it should either be configurable (and I hate adding settings) or you'll need to do two calls.

> If it's useless we should not add anything at all disguised_face

That was precisely my point; if it's useless when debugging or waiting for a task, let's push it somewhere else dedicated to that instead of returning it, where 99% of the time, it's useless.

**Reply by bidoubiwa** [reply]
> of useless stats in it

> If it's useless we should not add anything at all 🥸

I suppose that these stats may not be relevant as they provide info on stand-alone filters and not mixes of them.

Example:
My filter is: `{ indexUid: ["movies"], statuses: ["processed"] }`

The stats route provides me with the following information:
- indexes: 
    - `movies`: 5 tasks
    - `reviews`: 10 tasks
- statuses: 
   - `processed`: 10 tasks
   - etc..


How can I know how many `processed` `movies` tasks there are with the stats object? The provided stats are not related to my query, and only provides information on stand-alone filters. These might enough for some use-cases, but in the use use-case I want to use the task count, it is not useful.

**Comment by chayaline** [comment]
About the use cases, here is how it could be used on a UI, in this case, Meilisearch Cloud:
![tasks - type filter](https://github.com/meilisearch/product/assets/16258667/6fe9a52c-ab67-444b-a372-44fff9153abd)
![tasks - index filter](https://github.com/meilisearch/product/assets/16258667/412b5e07-98be-4f9c-abd0-5c0a979fe894)

We would fetch a number of tasks, here I put 200k, but it's not decided yet, then display them 50 by 50 with an automatic `show more` button that will show 50 more when you reach the bottom of your currently displayed list. 
The numbers would be updated according to the applied filters. For example: the default date filter is set to `last 24hrs`, all types, status, and indexes are shown, the filter shows: `document deletion (118)`. Now I filter to show only the failed tasks, the filter would now display `document deletion (20)`. If I change the date to `last 7 days` it would change to `document deletion (80)` etc...

I am not a developer anymore, but I agree that having to do 2 separate calls to get the tasks and the tasks stats seems not to be a good DX IMO.

**Reply by bidoubiwa** [reply]
If I understand correctly, you need a totalTask that reflects your filter, which can contain a multitude of different conditions. Which, if I understood correctly, is not what the `stats` object provides.

**Reply by chayaline** [reply]
Not really, it's the total number fetched, not taking into account the filtered number. But I wonder what's more useful, this or the actual total number of tasks in the project or the number of displayed tasks (reflecting the filters) 🤔

**Reply by bidoubiwa** [reply]
I'm not sure I understood what you mean with `total number fetched` ? The number of displayed tasks is probably `results.length` if I understood the difference between both.

**Reply by irevoire** [reply]
Hey, I'm not sure I understood the following sentence; sorry :sweat: 

> Not really, it's the total number fetched, not taking into account the filtered number.

Isn't the « total number fetched » 20 by default and set by the limit?
Or do you just want the total number of tasks in the task queue?

And the 
> number of displayed tasks (reflecting the filters)

Would be the total number of tasks in the task queue **after** applying the filters?

**Reply by chayaline** [reply]
I'm starting to get confused too 😅 
for me the displayed tasks are the ones actually shown on the UI. We decide to show them 50 by 50, but maybe we fetched 200 of them, so the `result.length` would be 200. Idk how that would work, gotta ask our front-end developers.

> Would be the total number of tasks in the task queue after applying the filters?

Yes, the tasks after applying filters. But isn't the queue only the `enqueued` and `ongoing task`? here we display all, including finished tasks

**Comment by bidoubiwa** [comment]
I had a chat with @chayaline to clarify their use case.

This is what their use-case needs:

## total tasks based on the filters.

For example, if I want all the tasks of movies from the last 24h (see image)

```json
{
    "indexUids": ["movies"],
    "afterEnqueuedAt": "now - 24h"
}
```

Which in UI translates to this:

<img width="1084" alt="Screenshot 2023-05-17 at 17 11 30" src="https://github.com/meilisearch/product/assets/33010418/9696acf0-361a-41a2-afbf-7cb21b5168ec">


In the upper right corner of the image, the `total tasks fetched` should represent the number of tasks that follow these conditions.
In this case `10`.

This number is easily provided by the `totalTasks` field returned by the `GET /tasks` route.

## Stats based on the filters

While totalTasks help for the upper right part of the screen, it does not help for the facet distribution of specific categories of tasks. For example in the following image:

<img width="1052" alt="Screenshot 2023-05-17 at 17 09 18" src="https://github.com/meilisearch/product/assets/33010418/10704b1d-91bc-4ce2-8a05-c3ccafc34eaf">


In this case the `stats` object really becomes interesting. But since it may be out of scope of this discussion, I opened a new one here:
https://github.com/meilisearch/product/discussions/654
